EIPS/eip-2537.md

```diff
@@ -0,0 +1,296 @@
+
+#### Prevention of DDoS on error handling
+
+This precompile performs extensive computations and in case of any errors during execution it MUST consume all gas from the the gas schedule for the corresponding operation.
+
+#### Gas schedule
+
+Assuming a constant `30 MGas/second` following prices are suggested.
+
+##### G1 addition
+
+`600` gas
+
+##### G1 multiplication
+
+`12000` gas
+
+##### G2 addition
+
+`4500` gas
+
+##### G2 multiplication
+
+`55000` gas
+
+##### G1/G2 Multiexponentiation
+
+Multiexponentiations are expected to be performed by the Peppinger algorithm (we can also say that is **must** be performed by Peppinger algorithm to have a speedup that results in a discount over naive implementation by multiplying each pair separately and adding the results). For this case there was a table prepared for discount in case of `k <= 128` points in the multiexponentiation with a discount cup `max_discount` for `k > 128`.
+
+To avoid non-integer arithmetic call cost is calculated as `k * multiplication_cost * discount / multiplier` where `multiplier = 1000`, `k` is a number of (scalar, point) pairs for the call, `multiplication_cost` is a corresponding single multiplication call cost for G1/G2.
+
+Discounts table as a vector of pairs `[k, discount]`:
+
+```
+[[1, 1200], [2, 888], [3, 764], [4, 641], [5, 594], [6, 547], [7, 500], [8, 453], [9, 438], [10, 423], [11, 408], [12, 394], [13, 379], [14, 364], [15, 349], [16, 334], [17, 330], [18, 326], [19, 322], [20, 318], [21, 314], [22, 310], [23, 306], [24, 302], [25, 298], [26, 294], [27, 289], [28, 285], [29, 281], [30, 277], [31, 273], [32, 269], [33, 268], [34, 266], [35, 265], [36, 263], [37, 262], [38, 260], [39, 259], [40, 257], [41, 256], [42, 254], [43, 253], [44, 251], [45, 250], [46, 248], [47, 247], [48, 245], [49, 244], [50, 242], [51, 241], [52, 239], [53, 238], [54, 236], [55, 235], [56, 233], [57, 232], [58, 231], [59, 229], [60, 228], [61, 226], [62, 225], [63, 223], [64, 222], [65, 221], [66, 220], [67, 219], [68, 219], [69, 218], [70, 217], [71, 216], [72, 216], [73, 215], [74, 214], [75, 213], [76, 213], [77, 212], [78, 211], [79, 211], [80, 210], [81, 209], [82, 208], [83, 208], [84, 207], [85, 206], [86, 205], [87, 205], [88, 204], [89, 203], [90, 202], [91, 202], [92, 201], [93, 200], [94, 199], [95, 199], [96, 198], [97, 197], [98, 196], [99, 196], [100, 195], [101, 194], [102, 193], [103, 193], [104, 192], [105, 191], [106, 191], [107, 190], [108, 189], [109, 188], [110, 188], [111, 187], [112, 186], [113, 185], [114, 185], [115, 184], [116, 183], [117, 182], [118, 182], [119, 181], [120, 180], [121, 179], [122, 179], [123, 178], [124, 177], [125, 176], [126, 176], [127, 175], [128, 174]]
+```
+
+`max_discount = 174`
+
+##### Pairing operaiton
+
+Cost of the pairing operation is `23000*k + 115000` where `k` is a number of pairs.
+
+##### Field-to-curve mappign operation
+
+Fp -> G1 mapping is `5500` gas.
+Fp2 -> G2 mapping is `110000` gas
+
```

Comment:
If `Fp` and `Fp2` are determined strictly by size of input then this should really be 2 precompiles.

Reply:
As Alex has commented in a spec above it's a simple routing, but one precompile reflects a single semantic underneath - map a base field element into the curve point over this field.

Reply:
But the function represents two disctinct types, G1 and a G2, resulting in a covariant return type.  Yes, you can essentially "cast" it but from a programming language perspective that means it's really two functions masquerading behind a if check.  I strongly feel these should be 2 different precompiles.

Reply:
I'm no mathematician. From a programming perspective, I would kind of agree with @shemnon about this. Would be interesting to hear from @ralexstokes or @vbuterin . 
Question: if this is used in e.g. a relay contract to verify the eth 2.0 destination address, would the it make any difference for the caller (contract) if it had to do the routing "externally" or whether it was done "internallly"? 
Would it always only use one of them for 2.0?  
cc @axic too

Reply:
i'm not that familiar with the underlying mathematics as well but if it is as simple as:

```python
def field_to_curve(some_element: Union[Fp,Fp2]) -> Union[G1Point, G2Point]:
    if is_fp(some_element):
        field_to_curve_fp(some_element)
    elif is_fp2(some_element):
        field_to_curve_fp2(some_element)
    else:
        raise Exception()
```

then i agree with having two separate precompiles -- another data point here is that we do it "this way" for all of the other operations; i.e. G1 and/or G2 variants of `add`, `mul`, etc. so then we would have G1 and G2 variants for the field to curve mapping

Reply:
@holiman as far as the eth2 cryptography is concerned, you end up with a particular configuration of using G1 or G2 in various places so that for the "eth2 verifier" we _will_ favor one over the other; in particular, we will want the G2 variant of this precompile and will not be using the G1 variant.

in those use cases where we want to call both/either, i don't see why pulling the conditional in the python i gave above into the EVM would be prohibitive.

Reply:
for concreteness, you can see an example here where we work over Fp2/G2 for the BLS sig verification: https://github.com/ralexstokes/deposit-contract-verifying-proxy/blob/master/deposit_contract_proxy.sol#L218-L233

Reply:
Unpopular opinion incoming.

@holiman if you ask me, honestly I would prefer a single precompile and not 9+ precompiles. I would prefer something like the following interface (https://gist.github.com/axic/8afa8ce7a262555e606668239287883c) in a single precompile.

Reason:
1) Lower overhead from contracts and we want to make sure these are used by contracts
2) If precompiles in the future would become contracts they wouldn't be 9 copies of the same thing (because we introduce a different VM which allows it or we introduce some other precompiles, e.g. montgomery multiplication, which would allow these written as libraries)

However, even if we have 9+ precompiles, I would still prefer to follow (a very restricted) ABI encoding. I've mentioned this on previous precompile discussions too.

Reply:
@axic G1 point is 48 bytes, not 16, and does not fit into any Ethereum type. In principle it's possible to declare it like
```
pragma solidity >=0.6.5 <0.7.0;

pragma experimental ABIEncoderV2;

interface Bls12Precompile {
    function g1_mul(uint256[2] calldata p0, uint256 scalar) external returns (uint256[2] memory);
    function pair(uint256[2][] calldata g1s, uint256[4][] calldata g2s) external returns (bool);
}
```

I'm not familiar if type aliasing can be made transparent using `struct G1 {uint256[2] inner}`. I don't have any opinion on it other than the fact that pulling ABI inside will be another level that needs to be tested.

Reply:
> G1 point is 48 bytes, not 16

It was late night and it seems I have mistaken "128 bytes" for "128 bits", consistently in the document.

Reply:
The interface was only an example, but it would be possible to define a very restricted encoding, which would be possible to be represented as an ABI at the same time.

However I think there's little chance for this to happen, given the opinion of others. We're pushing the marshalling complexity inside the EVM where contracts have to deal with it.